rclcpp/src/rclcpp/executor.cpp

```diff
@@ -138,95 +123,38 @@ Executor::~Executor()
 std::vector<rclcpp::CallbackGroup::WeakPtr>
 Executor::get_all_callback_groups()
 {
-  std::vector<rclcpp::CallbackGroup::WeakPtr> groups;
-  std::lock_guard<std::mutex> guard{mutex_};
-  for (const auto & group_node_ptr : weak_groups_associated_with_executor_to_nodes_) {
-    groups.push_back(group_node_ptr.first);
-  }
-  for (auto const & group_node_ptr : weak_groups_to_nodes_associated_with_executor_) {
-    groups.push_back(group_node_ptr.first);
-  }
-  return groups;
+  std::lock_guard<std::mutex> lock(mutex_);
+  return this->collector_.get_all_callback_groups();
```

Comment:
Should the various `Executor::get_all_callback_groups()` methods do a `this->collector_.update_collection()` ?

otherwise the following test would fail
```
auto executor = ...
auto node = ....
executor->add_node(node);
auto groups = executor->get_all_callback_groups();
```

here I would expect the `groups` to contain what I just added in the line before, but it won't happen with the current implementation.

I added the `update_collection` calls in the events-executor: https://github.com/ros2/rclcpp/pull/2155/files#diff-65295506f5cccb52eb107cb023966a999c0cc4076ccb92197b7c23d73ca7d767R360-R379

Reply:
I think it makes sense to do.  In the longer term, I wonder if these `get_` methods still make sense as part of the public executor API?  I imagine since the collector is now handling all of this, these are largely implementation details that can be eventually hidden.

In my early testing, we have to be cautious about manipulating the collection out from under the waitset, but I think now that I'm making a copy of the notify waitable for the duration of the wait, this shouldn't be an issue.

Reply:
Yes, I don't see any reason why users would need to access the callback groups from the executor.